Fix invitation location updates for new web users from bulk upload

mkangia opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
When bulk uploading web users, locations get removed from invites for new web users when the upload does not contain location column(s). In such a case, instead, no updates are expected for the location information on the invites.
This PR fixes that bug.

Additional bug was found for audit tracking of updates to invitations which were addressed as well.
In case the assigned locations and primary location were wiped out from invite (from UI or bulk upload), it was not being tracked in the audit. This was also fixed.

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-15959

In context of the method `create_or_update_web_user_invite` being modified, the changes are based on an agreement that if the value passed is None, the default value set by the method, then the calling method didn't intend for any changes for the respective field.
I am open to push back on this if this does not feel right. I have added some description in the commits if helpful.
I have seen the use of [Ellipsis](https://realpython.com/python-ellipsis/) at places, though I am not aligned with the idea of using it here.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The code area is well tested but was missing test for the case being considered, so I have added tests for that.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
I am going to run this through QA. This is a GA and the changes have a significant impact radius, though I will do that  once I have some feedback from those more familiar with this area of code.
https://dimagi.atlassian.net/browse/QA-8091

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
update when location codes passed

mkangia committed:
reserve None for no updates

- the value for primary_location_id and assigned_location_ids should be an empty string and an empty list respectively, when the values are to be removed

mkangia committed:
skip updates when None

- adapt method to when primary location id and assigned locations are passed as None
- No other functional change to the method

mkangia committed:
add tests

- test invite location removal
- test invite location information skipped

mkangia committed:
correct updates when removing locations from invite

mkangia commented:
Hey @Robert-Costello 
Do you mind sharing feedback on this? I see you worked on this feature last in https://github.com/dimagi/commcare-hq/pull/34550

fyi @esoergel feel free to chime in.

esoergel commented on corehq/apps/user_importer/importer.py:
```diff
@@ -332,44 +332,60 @@ def get_location_from_site_code(site_code, location_cache):
                                     assigned_location_ids=None, profile=None, tableau_role=None,
                                     tableau_group_ids=None, user_change_logger=None, send_email=True):
+    """
+    :param primary_location_id: use None for value to not be set or updated
```
Doesn't this mean that there's no way to remove a location assignment using this function?  Is that supported in the upload?  Like, if an upload file doesn't have a `primary_location_id` column, then I agree we probably shouldn't modify the value, but if it _does_ have one, and it's blank, should that mean "unassign"?

mkangia replied:
Hey  @esoergel 

>  Doesn't this mean that there's no way to remove a location assignment using this function?

An empty string for primary location and an empty list for assigned locations should remove the location assignments.

>  Like, if an upload file doesn't have a primary_location_id column, then I agree we probably shouldn't modify the value, but if it does have one, and it's blank, should that mean "unassign"?

Ya. If the column is left blank, then I expect that the user is asking for location assignment to be removed.
If they want no updates then the column isn't present.
I believe what someone would do is download the current structure, modify it and then upload it back. So, if they did blank out any column, I assume it would mean to remove it or else they would leave the value how it is.
I am not sure what is the expectation from product perspective across HQ for bulk uploads and I can cross check that with product team as a follow up if the changes do look okay code wise.

esoergel replied:
Nice, yeah that makes sense to me.  Not sure if we have a designated standard for this across product, but it does sound useful to have one.

esoergel commented on corehq/apps/user_importer/importer.py:
```diff
@@ -332,44 +332,60 @@ def get_location_from_site_code(site_code, location_cache):
-        invite_fields.pop('primary_location')
+
+        # pop already formatted fields before update
+        if invite_fields.get('primary_location'):
+            invite_fields.pop('primary_location')
```
Should this be
```suggestion
        invite_fields.pop('primary_location', None)
```
As-written, if `primary_location` is present and falsy, it'll stay.

gherceg commented on corehq/apps/user_importer/tests/test_importer.py:
```diff
@@ -2066,7 +2066,55 @@ def test_invite_location_add(self):
+        self.setup_locations()
+        import_users_and_groups(
+            self.domain.name,
+            [self._get_invited_spec(location_code=[a.site_code for a in [self.loc1]])],
```
nit: I know you are just keeping this how it was done previously, but feels unnecessarily complicated when using one location.
```suggestion
            [self._get_invited_spec(location_code=[self.loc1.site_code])],
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/36931/commits/b8215c25ff3a0d87610f6a29b69ae88c01bb5375

gherceg reviewed (commented):
Does it seem worth adding a test with multiple location codes since the logic for primary and assigned varies slightly?

mkangia commented:
Hey  @gherceg 

> Does it seem worth adding a test with multiple location codes since the logic for primary and assigned varies slightly?

It does. I will add that.

mkangia committed:
refactor

mkangia committed:
adapt tests to multiple locations

mkangia committed:
refactor

mkangia commented on corehq/apps/user_importer/importer.py:
```diff
@@ -332,44 +332,59 @@ def get_location_from_site_code(site_code, location_cache):
+    }
+
+    if primary_location_id is not None:
+        primary_location = SQLLocation.by_location_id(primary_location_id)
```
Thanks @esoergel 

Should we add that check on the model then? seems useful.

esoergel replied:
I was thinking here, but no strong preference

mkangia commented:
Hello @esoergel @gherceg 

So, I checked with product and we are aligned on the approach here. So we are all set here.
I will be submitting QA request but meanwhile if you can review this please.

mkangia committed:
Merge branch 'master' into mk/15959-fix-web-user-importer-invites

millerdev reviewed (commented):
> if the value passed is None, the default value set by the method, then the calling method didn't intend for any changes for the respective field.

This seems reasonable.

mkangia committed:
improve doc wording

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

mkangia committed:
fix bug & add tests for tracking invitation changes

mkangia commented on corehq/apps/users/views/__init__.py:
```diff
@@ -1437,11 +1437,11 @@ def format_changes(domain, changed_values):
             changed_values.update(UserChangeMessage.program_change(Program.get(program_id)))
-        assigned_locations = changed_values.pop('assigned_locations', None)
-        if assigned_locations:
+        if 'assigned_locations' in changed_values:
```
Earlier we were ignoring if assigned_locations was passed as an empty list, which would be in case the assigned locations are wiped out. So, the change here is to accept whatever was passed, if it was passed.

mkangia replied:
@esoergel just wanted to draw your attention to this specifically. `InvitationHistory` isn't exposed on the UI like UserHistory so this probably went unnoticed.

mkangia replied:
This PR was originally intended to fix the incorrect updates to location information on web user invites.
I am thinking if the fixes for Invitation History should be done separately or this is okay to be included here.
Open to suggestions.

esoergel replied:
I think that's probably fine to include here

mkangia replied:
> There are two other places where this method is called out from, other than the changes being made here.
One of them is an API end point for invitations and it appears to be having the same bug that this PR is trying to resolve with bulk upload

The API end point does not allow update but only create so this bug does not happen here. The bug was when locations were removed intentionally in an update. For a creation of an invite, if locations are assigned or not assigned, both will work okay.

mkangia committed:
expose invitation history on django admin

mkangia committed:
Merge branch 'master' into mk/15959-fix-web-user-importer-invites

mkangia commented on corehq/apps/user_importer/importer.py:
```diff
@@ -860,7 +860,7 @@ def process_new_user(self):
             user_invite_loc_id = None
             user_invite_locs_ids = []
-            if self.domain_info.can_assign_locations and cv['location_codes']:
+            if self.domain_info.can_assign_locations and cv['location_codes'] is not None:
```
note: This is the primary fix for the bug to be addressed.

This change here makes the logic same for the "new" user as it is done above for "existing user" in function `_modify_existing_user_in_domain`.

mkangia commented on corehq/apps/user_importer/importer.py:
```diff
@@ -338,40 +338,54 @@ def create_or_update_web_user_invite(email, domain, role_qualified_id, upload_us
+        'profile': profile
+    }
+
+    if primary_location_id is not None:
```
Note: Similar change done for primary location.

Additionally, adapt the function to `None` values for assigned locations and primary location.

mkangia committed:
Merge branch 'master' into mk/15959-fix-web-user-importer-invites

mkangia commented:
Hey @esoergel 

Thanks for your patience with this. I think I am good with this PR and would like to merge this now.

1. QA is complete and no issues were found.
2. I have tested this locally thoroughly so I am confident of the changes.
3. I have added some additional notes on the PR for better understanding of the commits.
4. Updated with latest master for another test run with updated code.

Requesting a final review/approval on this so I can go ahead with it.
